// @fuel-ts/providers/dist/index.js
"use strict";var Mt=Object.create;var V=Object.defineProperty;var At=Object.getOwnPropertyDescriptor;var Pt=Object.getOwnPropertyNames;var wt=Object.getPrototypeOf,Qt=Object.prototype.hasOwnProperty;var Dt=(a,e)=>{for(var t in e)V(a,t,{get:e[t],enumerable:!0})},pt=(a,e,t,r)=>{if(e&&typeof e=="object"||typeof e=="function")for(let n of Pt(e))!Qt.call(a,n)&&n!==t&&V(a,n,{get:()=>e[n],enumerable:!(r=At(e,n))||r.enumerable});return a};var lt=(a,e,t)=>(t=a!=null?Mt(wt(a)):{},pt(e||!a||!a.__esModule?V(t,"default",{value:a,enumerable:!0}):t,a)),Lt=a=>pt(V({},"__esModule",{value:!0}),a);var Ot=(a,e,t)=>{if(!e.has(a))throw TypeError("Cannot "+t)};var yt=(a,e,t)=>{if(e.has(a))throw TypeError("Cannot add the same private member more than once");e instanceof WeakSet?e.add(a):e.set(a,t)};var dt=(a,e,t)=>(Ot(a,e,"access private method"),t);var ge={};Dt(ge,{ChangeOutputCollisionError:()=>v,CoinStatus:()=>at,CreateTransactionRequest:()=>O,NoWitnessAtIndexError:()=>Z,NoWitnessByOwnerError:()=>z,Provider:()=>E,ScriptTransactionRequest:()=>R,TransactionResponse:()=>U,TransactionType:()=>u.TransactionType,calculatePriceWithFactor:()=>Q,coinQuantityfy:()=>k,getGasUsedFromReceipts:()=>_,getSignableTransaction:()=>xt,getTransactionId:()=>pe,inputify:()=>nt,outputify:()=>st,returnZeroScript:()=>ot,transactionRequestify:()=>$});module.exports=Lt(ge);var mt=require("@ethersproject/bytes"),et=require("@fuel-ts/constants"),gt=require("@fuel-ts/math"),k=a=>{var r,n;let e,t;return Array.isArray(a)?(t=a[0],e=(r=a[1])!=null?r:et.NativeAssetId):(t=a.amount,e=(n=a.assetId)!=null?n:et.NativeAssetId),{assetId:(0,mt.hexlify)(e),amount:(0,gt.bn)(t)}};var c=lt(require("graphql-tag")),at=(t=>(t.Spent="SPENT",t.Unspent="UNSPENT",t))(at||{});var M=c.default`
  fragment transactionFragment on Transaction {
    id
    rawPayload
    status {
      type: __typename
      ... on SubmittedStatus {
        time
      }
      ... on SuccessStatus {
        block {
          id
        }
        time
        programState {
          returnType
          data
        }
      }
      ... on FailureStatus {
        block {
          id
        }
        time
        reason
      }
    }
  }
`,St=c.default`
  fragment receiptFragment on Receipt {
    data
    rawPayload
  }
`,rt=c.default`
  fragment coinFragment on Coin {
    utxoId
    owner
    amount
    assetId
    maturity
    status
    blockCreated
  }
`,$t=c.default`
  fragment messageFragment on Message {
    amount
    sender
    recipient
    data
    nonce
    daHeight
  }
`,It=c.default`
  fragment balanceFragment on Balance {
    owner
    amount
    assetId
  }
`,H=c.default`
  fragment blockFragment on Block {
    id
    height
    producer
    transactions {
      id
    }
    time
  }
`,Ft=c.default`
  fragment chainInfoFragment on ChainInfo {
    name
    baseChainHeight
    peerCount
    consensusParameters {
      gasPriceFactor
      maxGasPerTx
      maxGasPerTx
      maxScriptLength
    }
    latestBlock {
      ...blockFragment
    }
  }
  ${H}
`,Ut=c.default`
  query getVersion {
    nodeInfo {
      nodeVersion
    }
  }
`,Et=c.default`
  query getInfo {
    nodeInfo {
      nodeVersion
      minGasPrice
    }
  }
`,Nt=c.default`
  query getChain {
    chain {
      ...chainInfoFragment
    }
  }
  ${Ft}
`,Vt=c.default`
  query getTransaction($transactionId: TransactionId!) {
    transaction(id: $transactionId) {
      ...transactionFragment
    }
  }
  ${M}
`,Ht=c.default`
  query getTransactionWithReceipts($transactionId: TransactionId!) {
    transaction(id: $transactionId) {
      ...transactionFragment
      receipts {
        ...receiptFragment
      }
    }
  }
  ${M}
  ${St}
`,Wt=c.default`
  query getTransactions($after: String, $before: String, $first: Int, $last: Int) {
    transactions(after: $after, before: $before, first: $first, last: $last) {
      edges {
        node {
          ...transactionFragment
        }
      }
    }
  }
  ${M}
`,Kt=c.default`
  query getTransactionsByOwner(
    $owner: Address!
    $after: String
    $before: String
    $first: Int
    $last: Int
  ) {
    transactionsByOwner(owner: $owner, after: $after, before: $before, first: $first, last: $last) {
      edges {
        node {
          ...transactionFragment
        }
      }
    }
  }
  ${M}
`,vt=c.default`
  query getBlock($blockId: BlockId, $blockHeight: U64) {
    block(id: $blockId, height: $blockHeight) {
      ...blockFragment
    }
  }
  ${H}
`,Zt=c.default`
  query getBlockWithTransactions($blockId: BlockId, $blockHeight: U64) {
    block(id: $blockId, height: $blockHeight) {
      ...blockFragment
      transactions {
        ...transactionFragment
      }
    }
  }
  ${H}
  ${M}
`,zt=c.default`
  query getBlocks($after: String, $before: String, $first: Int, $last: Int) {
    blocks(after: $after, before: $before, first: $first, last: $last) {
      edges {
        node {
          ...blockFragment
        }
      }
    }
  }
  ${H}
`,Xt=c.default`
  query getCoin($coinId: UtxoId!) {
    coin(utxoId: $coinId) {
      ...coinFragment
    }
  }
  ${rt}
`,Jt=c.default`
  query getCoins(
    $filter: CoinFilterInput!
    $after: String
    $before: String
    $first: Int
    $last: Int
  ) {
    coins(filter: $filter, after: $after, before: $before, first: $first, last: $last) {
      edges {
        node {
          ...coinFragment
        }
      }
    }
  }
  ${rt}
`,Yt=c.default`
  query getCoinsToSpend(
    $owner: Address!
    $spendQuery: [SpendQueryElementInput!]!
    $maxInputs: Int
    $excludedIds: [UtxoId!]
  ) {
    coinsToSpend(
      owner: $owner
      spendQuery: $spendQuery
      maxInputs: $maxInputs
      excludedIds: $excludedIds
    ) {
      ...coinFragment
    }
  }
  ${rt}
`,jt=c.default`
  query getContract($contractId: ContractId!) {
    contract(id: $contractId) {
      bytecode
      id
    }
  }
`,te=c.default`
  query getBalance($owner: Address!, $assetId: AssetId!) {
    balance(owner: $owner, assetId: $assetId) {
      ...balanceFragment
    }
  }
  ${It}
`,ee=c.default`
  query getBalances(
    $filter: BalanceFilterInput!
    $after: String
    $before: String
    $first: Int
    $last: Int
  ) {
    balances(filter: $filter, after: $after, before: $before, first: $first, last: $last) {
      edges {
        node {
          ...balanceFragment
        }
      }
    }
  }
  ${It}
`,ae=c.default`
  query getMessages($owner: Address!, $after: String, $before: String, $first: Int, $last: Int) {
    messages(owner: $owner, after: $after, before: $before, first: $first, last: $last) {
      edges {
        node {
          ...messageFragment
        }
      }
    }
  }
  ${$t}
`,re=c.default`
  mutation dryRun($encodedTransaction: HexString!, $utxoValidation: Boolean) {
    dryRun(tx: $encodedTransaction, utxoValidation: $utxoValidation) {
      ...receiptFragment
    }
  }
  ${St}
`,ne=c.default`
  mutation submit($encodedTransaction: HexString!) {
    submit(tx: $encodedTransaction) {
      id
    }
  }
`,se=c.default`
  mutation startSession {
    startSession
  }
`,oe=c.default`
  mutation endSession($sessionId: ID!) {
    endSession(id: $sessionId)
  }
`,ie=c.default`
  mutation execute($sessionId: ID!, $op: String!) {
    execute(id: $sessionId, op: $op)
  }
`,ce=c.default`
  mutation reset($sessionId: ID!) {
    reset(id: $sessionId)
  }
`,ue=(a,e,t)=>a();function bt(a,e=ue){return{getVersion(t,r){return e(n=>a.request(Ut,t,{...r,...n}),"getVersion","query")},getInfo(t,r){return e(n=>a.request(Et,t,{...r,...n}),"getInfo","query")},getChain(t,r){return e(n=>a.request(Nt,t,{...r,...n}),"getChain","query")},getTransaction(t,r){return e(n=>a.request(Vt,t,{...r,...n}),"getTransaction","query")},getTransactionWithReceipts(t,r){return e(n=>a.request(Ht,t,{...r,...n}),"getTransactionWithReceipts","query")},getTransactions(t,r){return e(n=>a.request(Wt,t,{...r,...n}),"getTransactions","query")},getTransactionsByOwner(t,r){return e(n=>a.request(Kt,t,{...r,...n}),"getTransactionsByOwner","query")},getBlock(t,r){return e(n=>a.request(vt,t,{...r,...n}),"getBlock","query")},getBlockWithTransactions(t,r){return e(n=>a.request(Zt,t,{...r,...n}),"getBlockWithTransactions","query")},getBlocks(t,r){return e(n=>a.request(zt,t,{...r,...n}),"getBlocks","query")},getCoin(t,r){return e(n=>a.request(Xt,t,{...r,...n}),"getCoin","query")},getCoins(t,r){return e(n=>a.request(Jt,t,{...r,...n}),"getCoins","query")},getCoinsToSpend(t,r){return e(n=>a.request(Yt,t,{...r,...n}),"getCoinsToSpend","query")},getContract(t,r){return e(n=>a.request(jt,t,{...r,...n}),"getContract","query")},getBalance(t,r){return e(n=>a.request(te,t,{...r,...n}),"getBalance","query")},getBalances(t,r){return e(n=>a.request(ee,t,{...r,...n}),"getBalances","query")},getMessages(t,r){return e(n=>a.request(ae,t,{...r,...n}),"getMessages","query")},dryRun(t,r){return e(n=>a.request(re,t,{...r,...n}),"dryRun","mutation")},submit(t,r){return e(n=>a.request(ne,t,{...r,...n}),"submit","mutation")},startSession(t,r){return e(n=>a.request(se,t,{...r,...n}),"startSession","mutation")},endSession(t,r){return e(n=>a.request(oe,t,{...r,...n}),"endSession","mutation")},execute(t,r){return e(n=>a.request(ie,t,{...r,...n}),"execute","mutation")},reset(t,r){return e(n=>a.request(ce,t,{...r,...n}),"reset","mutation")}}}var b=require("@ethersproject/bytes"),Rt=require("@fuel-ts/abi-coder"),it=require("@fuel-ts/address"),ct=require("@fuel-ts/constants"),i=require("@fuel-ts/math"),y=require("@fuel-ts/transactions"),_t=require("graphql-request"),kt=lt(require("lodash.clonedeep"));var l=require("@ethersproject/bytes"),W=require("@fuel-ts/constants"),f=require("@fuel-ts/math"),h=require("@fuel-ts/transactions"),nt=a=>{var e,t,r,n,s;switch(a.type){case h.InputType.Coin:{let o=(0,l.arrayify)((e=a.predicate)!=null?e:"0x"),p=(0,l.arrayify)((t=a.predicateData)!=null?t:"0x");return{type:h.InputType.Coin,utxoID:{transactionId:(0,l.hexlify)((0,l.arrayify)(a.id).slice(0,32)),outputIndex:(0,l.arrayify)(a.id)[32]},owner:(0,l.hexlify)(a.owner),amount:(0,f.bn)(a.amount),assetId:(0,l.hexlify)(a.assetId),txPointer:{blockHeight:(0,f.toNumber)((0,l.arrayify)(a.txPointer).slice(0,8)),txIndex:(0,f.toNumber)((0,l.arrayify)(a.txPointer).slice(8,16))},witnessIndex:a.witnessIndex,maturity:(r=a.maturity)!=null?r:0,predicateLength:o.length,predicateDataLength:p.length,predicate:(0,l.hexlify)(o),predicateData:(0,l.hexlify)(p)}}case h.InputType.Contract:return{type:h.InputType.Contract,utxoID:{transactionId:W.ZeroBytes32,outputIndex:0},balanceRoot:W.ZeroBytes32,stateRoot:W.ZeroBytes32,txPointer:{blockHeight:(0,f.toNumber)((0,l.arrayify)(a.txPointer).slice(0,8)),txIndex:(0,f.toNumber)((0,l.arrayify)(a.txPointer).slice(8,16))},contractID:(0,l.hexlify)(a.contractId)};case h.InputType.Message:{let o=(0,l.arrayify)((n=a.predicate)!=null?n:"0x"),p=(0,l.arrayify)((s=a.predicateData)!=null?s:"0x");return{type:h.InputType.Message,sender:(0,l.hexlify)(a.sender),recipient:(0,l.hexlify)(a.recipient),amount:(0,f.bn)(a.amount),nonce:(0,f.bn)(a.nonce),witnessIndex:a.witnessIndex,dataLength:a.data.length,predicateLength:o.length,predicateDataLength:p.length,data:a.data,predicate:(0,l.hexlify)(o),predicateData:(0,l.hexlify)(p)}}default:throw new Error("Invalid Input type")}};var G=require("@ethersproject/bytes"),A=require("@fuel-ts/constants"),P=require("@fuel-ts/math"),S=require("@fuel-ts/transactions"),st=a=>{switch(a.type){case S.OutputType.Coin:return{type:S.OutputType.Coin,to:(0,G.hexlify)(a.to),amount:(0,P.bn)(a.amount),assetId:(0,G.hexlify)(a.assetId)};case S.OutputType.Contract:return{type:S.OutputType.Contract,inputIndex:a.inputIndex,balanceRoot:A.ZeroBytes32,stateRoot:A.ZeroBytes32};case S.OutputType.Withdrawal:return{type:S.OutputType.Withdrawal,to:(0,G.hexlify)(a.to),amount:(0,P.bn)(a.amount),assetId:(0,G.hexlify)(a.assetId)};case S.OutputType.Change:return{type:S.OutputType.Change,to:(0,G.hexlify)(a.to),amount:(0,P.bn)(0),assetId:(0,G.hexlify)(a.assetId)};case S.OutputType.Variable:return{type:S.OutputType.Variable,to:A.ZeroBytes32,amount:(0,P.bn)(0),assetId:A.ZeroBytes32};case S.OutputType.ContractCreated:return{type:S.OutputType.ContractCreated,contractId:(0,G.hexlify)(a.contractId),stateRoot:(0,G.hexlify)(a.stateRoot)};default:throw new Error("Invalid Output type")}};var g=require("@ethersproject/bytes"),C=require("@fuel-ts/address"),B=require("@fuel-ts/constants"),L=require("@fuel-ts/math"),u=require("@fuel-ts/transactions");var qt=require("@ethersproject/sha2"),T=require("@fuel-ts/constants"),w=require("@fuel-ts/math"),m=require("@fuel-ts/transactions"),xt=a=>{let e={...a};switch(e.type){case m.TransactionType.Script:{e.receiptsRoot=T.ZeroBytes32;break}case m.TransactionType.Create:break;default:throw new Error("Not implemented")}return e.inputs=e.inputs.map(t=>t.type===m.InputType.Contract?{...t,utxoID:{transactionId:T.ZeroBytes32,outputIndex:0},balanceRoot:T.ZeroBytes32,stateRoot:T.ZeroBytes32}:t),e.outputs=e.outputs.map(t=>{switch(t.type){case m.OutputType.Contract:return{...t,balanceRoot:T.ZeroBytes32,stateRoot:T.ZeroBytes32};case m.OutputType.Change:return{...t,amount:(0,w.bn)(0)};case m.OutputType.Variable:return{...t,to:T.ZeroBytes32,amount:(0,w.bn)(0),assetId:T.ZeroBytes32};default:return t}}),e},pe=a=>{let e=xt(a),t=new m.TransactionCoder().encode(e);return(0,qt.sha256)(t)},Q=(a,e,t)=>(0,w.bn)(Math.ceil(a.toNumber()/t.toNumber())*e.toNumber()),_=a=>{let e=a.find(t=>t.type===m.ReceiptType.ScriptResult);return e&&e.type===m.ReceiptType.ScriptResult?e.gasUsed:(0,w.bn)(0)};var D=require("@ethersproject/bytes"),le=a=>{let e=new Uint8Array(32);return e.set((0,D.arrayify)(a)),e},ft=a=>{let e,t;return Array.isArray(a)?(e=a[0],t=a[1]):(e=a.key,t=a.value),{key:(0,D.hexlify)(e),value:(0,D.hexlify)(le(t))}};var K=require("@ethersproject/bytes"),Gt=a=>{let e=(0,K.arrayify)(a);return{data:(0,K.hexlify)(e),dataLength:e.length}};var ot={bytes:(0,g.arrayify)("0x24000000"),encodeScriptData:()=>new Uint8Array(0)},v=class extends Error{constructor(){super(...arguments);this.name="ChangeOutputCollisionError";this.message='A ChangeOutput with the same "assetId" already exists for a different "to" address'}},Z=class extends Error{constructor(t){super();this.index=t;this.name="NoWitnessAtIndexError";this.message=`Witness at index "${t}" was not found`}},z=class extends Error{constructor(t){super();this.owner=t;this.name="NoWitnessByOwnerError";this.message=`A witness for the given owner "${t}" was not found`}},X=class{constructor({gasPrice:e,gasLimit:t,maturity:r,inputs:n,outputs:s,witnesses:o}={}){this.inputs=[];this.outputs=[];this.witnesses=[];this.gasPrice=(0,L.bn)(e!=null?e:0),this.gasLimit=(0,L.bn)(t!=null?t:0),this.maturity=r!=null?r:0,this.inputs=[...n!=null?n:[]],this.outputs=[...s!=null?s:[]],this.witnesses=[...o!=null?o:[]]}getBaseTransaction(){var n,s,o,p,x,q;let e=(s=(n=this.inputs)==null?void 0:n.map(nt))!=null?s:[],t=(p=(o=this.outputs)==null?void 0:o.map(st))!=null?p:[],r=(q=(x=this.witnesses)==null?void 0:x.map(Gt))!=null?q:[];return{gasPrice:this.gasPrice,gasLimit:this.gasLimit,maturity:this.maturity,inputs:e,outputs:t,witnesses:r,inputsCount:e.length,outputsCount:t.length,witnessesCount:r.length}}toTransactionBytes(){return new u.TransactionCoder().encode(this.toTransaction())}pushInput(e){return this.inputs.push(e),this.inputs.length-1}pushOutput(e){return this.outputs.push(e),this.outputs.length-1}createWitness(){return this.witnesses.push("0x"),this.witnesses.length-1}updateWitness(e,t){if(!this.witnesses[e])throw new Z(e);this.witnesses[e]=t}getCoinInputs(){return this.inputs.filter(e=>e.type===u.InputType.Coin)}getCoinOutputs(){return this.outputs.filter(e=>e.type===u.OutputType.Coin)}getChangeOutputs(){return this.outputs.filter(e=>e.type===u.OutputType.Change)}getCoinInputWitnessIndexByOwner(e){var r,n;let t=(0,C.addressify)(e);return(n=(r=this.inputs.find(s=>s.type===u.InputType.Coin&&s.owner===t.toB256()))==null?void 0:r.witnessIndex)!=null?n:null}updateWitnessByCoinInputOwner(e,t){let r=this.getCoinInputWitnessIndexByOwner(e);if(!r)throw new z((0,C.addressify)(e));this.updateWitness(r,t)}addCoin(e){let t=this.getCoinInputWitnessIndexByOwner(C.Address.fromB256(e.owner));typeof t!="number"&&(t=this.createWitness()),this.pushInput({type:u.InputType.Coin,...e,witnessIndex:t,txPointer:"0x00000000000000000000000000000000"});let r=this.getChangeOutputs().find(n=>(0,g.hexlify)(n.assetId)===e.assetId);if(r&&(0,g.hexlify)(r.to)!==e.owner)throw new v;r||this.pushOutput({type:u.OutputType.Change,to:e.owner,assetId:e.assetId})}addCoins(e){e.forEach(t=>this.addCoin(t))}addCoinOutput(e,t,r=B.NativeAssetId){this.pushOutput({type:u.OutputType.Coin,to:(0,C.addressify)(e).toB256(),amount:t,assetId:r})}addCoinOutputs(e,t){t.map(k).forEach(r=>{this.pushOutput({type:u.OutputType.Coin,to:(0,C.addressify)(e).toB256(),amount:r.amount,assetId:r.assetId})})}byteSize(){return this.toTransactionBytes().length}chargeableByteSize(){let e=this.witnesses.reduce((t,r)=>t+(0,g.arrayify)(r).length,0);return(0,L.bn)(this.toTransactionBytes().length-e)}calculateFee(){let e=Q(this.gasLimit,this.gasPrice,u.GAS_PRICE_FACTOR);return{assetId:B.NativeAssetId,amount:e.isZero()?(0,L.bn)(1):e}}addMessage(e){let t=this.getCoinInputWitnessIndexByOwner(e.recipient);typeof t!="number"&&(t=this.createWitness()),this.pushInput({type:u.InputType.Message,...e,sender:e.sender.toBytes(),recipient:e.recipient.toBytes(),witnessIndex:t})}addMessages(e){e.forEach(t=>this.addMessage(t))}},R=class extends X{constructor({script:t,scriptData:r,...n}={}){super(n);this.type=u.TransactionType.Script;this.script=(0,g.arrayify)(t!=null?t:ot.bytes),this.scriptData=(0,g.arrayify)(r!=null?r:ot.encodeScriptData())}static from(t){return t instanceof this?t:new this(t)}toTransaction(){var n,s;let t=(0,g.arrayify)((n=this.script)!=null?n:"0x"),r=(0,g.arrayify)((s=this.scriptData)!=null?s:"0x");return{type:u.TransactionType.Script,...super.getBaseTransaction(),scriptLength:t.length,scriptDataLength:r.length,receiptsRoot:B.ZeroBytes32,script:(0,g.hexlify)(t),scriptData:(0,g.hexlify)(r)}}getContractInputs(){return this.inputs.filter(t=>t.type===u.InputType.Contract)}getContractOutputs(){return this.outputs.filter(t=>t.type===u.OutputType.Contract)}getVariableOutputs(){return this.outputs.filter(t=>t.type===u.OutputType.Variable)}setScript(t,r){this.script=t.bytes,this.scriptData=t.encodeScriptData(r),this.bytesOffset===void 0&&(this.bytesOffset=this.scriptData.byteLength)}addVariableOutputs(t=1){let r=t;for(;r;)this.pushOutput({type:u.OutputType.Variable}),r-=1;return this.outputs.length-1}addContract(t){let r=(0,C.addressify)(t);if(this.getContractInputs().find(s=>s.contractId===r.toB256()))return;let n=super.pushInput({type:u.InputType.Contract,contractId:r.toB256(),txPointer:"0x00000000000000000000000000000000"});this.pushOutput({type:u.OutputType.Contract,inputIndex:n})}},O=class extends X{constructor({bytecodeWitnessIndex:t,salt:r,storageSlots:n,...s}={}){super(s);this.type=u.TransactionType.Create;this.bytecodeWitnessIndex=t!=null?t:0,this.salt=(0,g.hexlify)(r!=null?r:B.ZeroBytes32),this.storageSlots=[...n!=null?n:[]]}static from(t){return t instanceof this?t:new this(t)}toTransaction(){var s,o;let t=this.getBaseTransaction(),r=this.bytecodeWitnessIndex,n=(o=(s=this.storageSlots)==null?void 0:s.map(ft))!=null?o:[];return{type:u.TransactionType.Create,...t,bytecodeLength:t.witnesses[r].dataLength/4,bytecodeWitnessIndex:r,storageSlotsCount:n.length,salt:this.salt?(0,g.hexlify)(this.salt):B.ZeroBytes32,storageSlots:n}}getContractCreatedOutputs(){return this.outputs.filter(t=>t.type===u.OutputType.ContractCreated)}addContractCreatedOutput(t,r){this.pushOutput({type:u.OutputType.ContractCreated,contractId:t,stateRoot:r})}},$=a=>{if(a instanceof R||a instanceof O)return a;switch(a.type){case u.TransactionType.Script:return R.from(a);case u.TransactionType.Create:return O.from(a);default:throw new Error(`Unknown transaction type: ${a.type}`)}};var Ct=require("@ethersproject/bytes"),ht=require("@fuel-ts/math"),F=require("@fuel-ts/transactions");var Tt=a=>{let e=new F.ReceiptCoder().decode((0,Ct.arrayify)(a.rawPayload),0)[0];switch(e.type){case F.ReceiptType.ReturnData:return{...e,data:a.data};case F.ReceiptType.LogData:return{...e,data:a.data};default:return e}},J,Bt,U=class{constructor(e,t,r){yt(this,J);this.gasUsed=(0,ht.bn)(0);this.id=e,this.request=t,this.provider=r}async waitForResult(){var t;let e=await dt(this,J,Bt).call(this);switch((t=e.status)==null?void 0:t.type){case"SubmittedStatus":throw new Error("Not yet implemented");case"FailureStatus":{let r=e.receipts.map(Tt);return this.gasUsed=_(r),{status:{type:"failure",reason:e.status.reason},receipts:r,transactionId:this.id,blockId:e.status.block.id,time:e.status.time}}case"SuccessStatus":{let r=e.receipts.map(Tt);return this.gasUsed=_(r),{status:{type:"success",programState:e.status.programState},receipts:r,transactionId:this.id,blockId:e.status.block.id,time:e.status.time}}default:throw new Error("Invalid Transaction status")}}async wait(){let e=await this.waitForResult();if(e.status.type==="failure")throw new Error(`Transaction failed: ${e.status.reason}`);return e}};J=new WeakSet,Bt=async function(){let{transaction:e}=await this.provider.operations.getTransactionWithReceipts({transactionId:this.id});if(!e)throw new Error("No Transaction was received from the client.");return e};var ye=a=>{let e=new y.ReceiptCoder().decode((0,b.arrayify)(a.rawPayload),0)[0];switch(e.type){case y.ReceiptType.ReturnData:return{...e,data:a.data};case y.ReceiptType.LogData:return{...e,data:a.data};default:return e}},de=a=>({name:a.name,baseChainHeight:(0,i.bn)(a.baseChainHeight),peerCount:a.peerCount,consensusParameters:{gasPriceFactor:(0,i.bn)(a.consensusParameters.gasPriceFactor),maxGasPerTx:(0,i.bn)(a.consensusParameters.maxGasPerTx),maxScriptLength:(0,i.bn)(a.consensusParameters.maxScriptLength)},latestBlock:{id:a.latestBlock.id,height:(0,i.bn)(a.latestBlock.height),producer:a.latestBlock.producer,time:a.latestBlock.time,transactions:a.latestBlock.transactions.map(e=>({id:e.id}))}}),me=a=>({minGasPrice:(0,i.bn)(a.minGasPrice),nodeVersion:a.nodeVersion}),E=class{constructor(e){this.url=e;let t=new _t.GraphQLClient(e);this.operations=bt(t)}async getVersion(){let{nodeInfo:{nodeVersion:e}}=await this.operations.getVersion();return e}async getNetwork(){return{name:"fuelv2",chainId:3735928559}}async getBlockNumber(){let{chain:e}=await this.operations.getChain();return(0,i.bn)(e.latestBlock.height,10)}async getNodeInfo(){let{nodeInfo:e}=await this.operations.getInfo();return me(e)}async getChain(){let{chain:e}=await this.operations.getChain();return de(e)}async sendTransaction(e){let t=$(e),r=(0,b.hexlify)(t.toTransactionBytes()),{gasUsed:n,minGasPrice:s}=await this.getTransactionCost(t,0);if((0,i.bn)(n).gt((0,i.bn)(t.gasLimit)))throw new Error(`gasLimit(${t.gasLimit}) is lower than the required (${n})`);if((0,i.bn)(s).gt((0,i.bn)(t.gasPrice)))throw new Error(`gasPrice(${t.gasPrice}) is lower than the required ${s}`);let{submit:{id:o}}=await this.operations.submit({encodedTransaction:r});return new U(o,t,this)}async call(e,{utxoValidation:t}={}){let r=$(e),n=(0,b.hexlify)(r.toTransactionBytes()),{dryRun:s}=await this.operations.dryRun({encodedTransaction:n,utxoValidation:t||!1});return{receipts:s.map(ye)}}async getTransactionCost(e,t=.2){let r=$((0,kt.default)(e)),{minGasPrice:n}=await this.getNodeInfo(),s=(0,i.max)(r.gasPrice,n),o=1+t;r.gasLimit=y.MAX_GAS_PER_TX,r.gasPrice=(0,i.bn)(0);let{receipts:p}=await this.call(r),x=(0,i.multiply)(_(p),o),q=Q(x,s,y.GAS_PRICE_FACTOR);return{minGasPrice:n,gasPrice:s,gasUsed:x,fee:q}}async getCoins(e,t,r){return(await this.operations.getCoins({first:10,...r,filter:{owner:e.toB256(),assetId:t&&(0,b.hexlify)(t)}})).coins.edges.map(o=>o.node).map(o=>({id:o.utxoId,assetId:o.assetId,amount:(0,i.bn)(o.amount),owner:o.owner,status:o.status,maturity:(0,i.bn)(o.maturity).toNumber(),blockCreated:(0,i.bn)(o.blockCreated)}))}async getCoinsToSpend(e,t,r,n){return(await this.operations.getCoinsToSpend({owner:e.toB256(),spendQuery:t.map(k).map(p=>({assetId:(0,b.hexlify)(p.assetId),amount:p.amount.toString(10)})),maxInputs:r,excludedIds:n==null?void 0:n.map(p=>(0,b.hexlify)(p))})).coinsToSpend.map(p=>({id:p.utxoId,status:p.status,assetId:p.assetId,amount:(0,i.bn)(p.amount),owner:p.owner,maturity:(0,i.bn)(p.maturity).toNumber(),blockCreated:(0,i.bn)(p.blockCreated)}))}async getBlock(e){let t;typeof e=="number"?t={blockHeight:(0,i.bn)(e).toString(10)}:e==="latest"?t={blockHeight:(await this.getBlockNumber()).toString(10)}:t={blockId:(0,i.bn)(e).toString(10)};let{block:r}=await this.operations.getBlock(t);return r?{id:r.id,height:(0,i.bn)(r.height),time:r.time,producer:r.producer,transactionIds:r.transactions.map(n=>n.id)}:null}async getBlockWithTransactions(e){let t;typeof e=="number"?t={blockHeight:(0,i.bn)(e).toString(10)}:e==="latest"?t={blockHeight:(await this.getBlockNumber()).toString()}:t={blockId:e};let{block:r}=await this.operations.getBlockWithTransactions(t);return r?{id:r.id,height:(0,i.bn)(r.height,10),time:r.time,producer:r.producer,transactionIds:r.transactions.map(n=>n.id),transactions:r.transactions.map(n=>{var s;return(s=new y.TransactionCoder().decode((0,b.arrayify)(n.rawPayload),0))==null?void 0:s[0]})}:null}async getTransaction(e){var r;let{transaction:t}=await this.operations.getTransaction({transactionId:e});return t?(r=new y.TransactionCoder().decode((0,b.arrayify)(t.rawPayload),0))==null?void 0:r[0]:null}async getContract(e){let{contract:t}=await this.operations.getContract({contractId:e});return t||null}async getBalance(e,t){let{balance:r}=await this.operations.getBalance({owner:e.toB256(),assetId:(0,b.hexlify)(t)});return(0,i.bn)(r.amount,10)}async getBalances(e,t){return(await this.operations.getBalances({first:10,...t,filter:{owner:e.toB256()}})).balances.edges.map(s=>s.node).map(s=>({assetId:s.assetId,amount:(0,i.bn)(s.amount)}))}async getMessages(e,t){return(await this.operations.getMessages({first:10,...t,owner:e.toB256()})).messages.edges.map(s=>s.node).map(s=>({amount:(0,i.bn)(s.amount),sender:it.Address.fromAddressOrString(s.sender),recipient:it.Address.fromAddressOrString(s.recipient),data:y.InputMessageCoder.decodeData(s.data),daHeight:(0,i.bn)(s.daHeight),nonce:(0,i.bn)(s.nonce)}))}async buildSpendPredicate(e,t,r,n,s=ct.NativeAssetId,o,p){let x=await this.getCoinsToSpend(e.address,[[t,s]]),q={fundTransaction:!0,...o},I=new R({gasLimit:y.MAX_GAS_PER_TX,...q}),Y;n&&e.types&&(Y=new Rt.AbiCoder().encode(e.types,n));let j=x.reduce((N,ut)=>(I.addCoin({...ut,predicate:e.bytes,predicateData:Y}),I.outputs=[],N.add(ut.amount)),(0,i.bn)(0));I.addCoinOutput(r,j,s);let tt=[];if(q.fundTransaction&&tt.push(I.calculateFee()),tt.length&&p){let N=await this.getCoinsToSpend(p,tt);I.addCoins(N)}return I}async submitSpendPredicate(e,t,r,n,s=ct.NativeAssetId,o,p){var q;let x=await this.buildSpendPredicate(e,t,r,n,s,o,p);try{return await(await this.sendTransaction(x)).waitForResult()}catch(I){throw(((q=I==null?void 0:I.response)==null?void 0:q.errors)||[]).some(({message:j})=>j.includes("unexpected block execution error TransactionValidity(InvalidPredicate"))?new Error("Invalid Predicate"):I}}};0&&(module.exports={ChangeOutputCollisionError,CoinStatus,CreateTransactionRequest,NoWitnessAtIndexError,NoWitnessByOwnerError,Provider,ScriptTransactionRequest,TransactionResponse,TransactionType,calculatePriceWithFactor,coinQuantityfy,getGasUsedFromReceipts,getSignableTransaction,getTransactionId,inputify,outputify,returnZeroScript,transactionRequestify});
